### PR TITLE
Notebook instance - Handling update only parameters

### PR DESCRIPTION
**Pull request by ananth102**
This pull request should be merged after #61, it contains code for handling cases where (lifecycleConfigName, AcceleratorTypes, DefaultCodeRepository, AdditionalCodeRepositories) are nil in desired and not nil in latest. In other words it allows the user to delete (lifecycleConfigName, AcceleratorTypes, DefaultCodeRepository, AdditionalCodeRepositories).

TODO: In test/e2e/tests/test_notebook_instance.py, we should use replace_custom_resource when test-infra supports it.

Files changed from #61 :
**generator.yaml:** Added a sdk_update_post_build_request hook that calls the handleUpdateOnlyParameters function
**pkg/resource/notebook_instance/sdk.go:** Added code that calls handleUpdateOnlyParameters
**test/e2e/tests/test_notebook_instance.py:** Added a test for Update only parameters.
**test/e2e/resources/notebook_instance.yaml:** Added the a field to test the update only parameters.

**pkg/resource/notebook_instance/custom_set_input_update.go:** Contains the function handleUpdateOnlyParameters which checks if the (lifecycleConfigName, AcceleratorTypes, DefaultCodeRepository, AdditionalCodeRepositories) are nil, if they are nil in the desired resource and not nil in the latest resource it sets the respective Disassociate<field> to true for the update input.

### CODE REVIEW

**Comment by mbaijal** [comment]
> TODO: In test/e2e/tests/test_notebook_instance.py, we should use replace_custom_resource when test-infra supports it.

Is there an issue already tracking this request ?

File: pkg/resource/notebook_instance/hooks.go [hunk 1]
```diff
@@ -48,3 +52,33 @@ func (rm *resourceManager) customSetOutput(r *resource) {
 	latestStatus := r.ko.Status.NotebookInstanceStatus
 	svccommon.SetSyncedCondition(r, latestStatus, &resourceName, &modifyingStatuses)
 }
+
+// CustomSetOutputDescribe has two functions
```
Comment (mbaijal) [inline pkg/resource/notebook_instance/hooks.go:56]: nit: three functions

File: pkg/resource/notebook_instance/hooks.go [hunk 2]
```diff
@@ -48,3 +52,33 @@ func (rm *resourceManager) customSetOutput(r *resource) {
...
+// 1. Setting the resource synced condition during sdk.Find
+// 2. Starting the Notebook if the StoppedByController field is set to UpdateTriggerd
+// 3. Reseting the StoppedByController field.
+func (rm *resourceManager) customSetOutputDescribe(r *resource,
+	ko *svcapitypes.NotebookInstance) error {
+	notebookStatus := *ko.Status.NotebookInstanceStatus // Get the Notebook State
+	if notebookStatus == svcsdk.NotebookInstanceStatusStopped {
+		if ko.Status.StoppedByControllerMETA != nil && *ko.Status.StoppedByControllerMETA == "UpdateTriggered" {
```
Comment (mbaijal) [inline pkg/resource/notebook_instance/hooks.go:64]: if I remember correctly we decided on the name `StoppedByControllerMetadata`, @surajkota ?

File: pkg/resource/notebook_instance/hooks.go [hunk 3]
```diff
@@ -48,3 +52,33 @@ func (rm *resourceManager) customSetOutput(r *resource) {
...
+			err := rm.startNotebookInstance(r)
+			if err != nil {
+				//We dont update StoppedByController here because we want it to try to start it again on the next reqeue.
```
Comment (mbaijal) [inline pkg/resource/notebook_instance/hooks.go:67]: Have you been able to test that this is working as expected ? Do you have a negative test case for updates ?

Comment (mbaijal) [reply pkg/resource/notebook_instance/hooks.go:67]: nit: use correct variable name in comments too.

File: pkg/resource/notebook_instance/hooks.go [hunk 4]
```diff
@@ -48,3 +52,33 @@ func (rm *resourceManager) customSetOutput(r *resource) {
...
+				return err
+			}
+			ko.Status.StoppedByControllerMETA = nil
+		}
+	}
+	rm.customSetOutput(&resource{ko}) // We set the sync status here
```
Comment (mbaijal) [inline pkg/resource/notebook_instance/hooks.go:73]: minor nit: avoid using pronouns in comments. I would recommend perusing through some comments on the rest of the repos in ACK and follow a similar pattern.

File: pkg/resource/notebook_instance/hooks.go [hunk 5]
```diff
@@ -48,3 +52,33 @@ func (rm *resourceManager) customSetOutput(r *resource) {
...
+	return nil
+}
+
+// customSetOutputUpdate sets the StoppedByController field to UpdateTriggered if it was in UpdatePending before.
```
Comment (mbaijal) [inline pkg/resource/notebook_instance/hooks.go:77]: nit: field name in comment

File: pkg/resource/notebook_instance/hooks.go [hunk 6]
```diff
@@ -48,3 +52,33 @@ func (rm *resourceManager) customSetOutput(r *resource) {
...
+func (rm *resourceManager) customSetOutputUpdate(ko *svcapitypes.NotebookInstance, latest *resource) {
+	//TODO: Replace the IsUpdating status with an annotation if the runtime can update annotations after a readOne call.
```
Comment (mbaijal) [inline pkg/resource/notebook_instance/hooks.go:79]: `isUpdating` is not in use anymore. However, the new field we added could also be converted into an annotation later.

File: generator.yaml [hunk 7]
```diff
@@ -642,14 +642,22 @@ resources:
...
+        template_path: notebook_instance/sdk_read_one_pre_set_output.go.tpl
+      sdk_read_one_post_set_output: 
+        template_path: notebook_instance/sdk_read_one_post_set_output.go.tpl
+      delta_pre_compare:
+        code: customSetDefaults(a, b)
+      sdk_update_pre_build_request:
+        template_path: notebook_instance/sdk_update_pre_build_request.go.tpl
+      sdk_update_post_build_request:
+        code: handleUpdateOnlyParameters(desired, latest, input)
+      sdk_update_post_set_output:
```
Comment (mbaijal) [inline generator.yaml:655]: this is only called after a successful update operation, correct ?

File: pkg/resource/notebook_instance/custom_set_delta.go [hunk 8]
```diff
@@ -4,8 +4,7 @@ import (
 	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
 )
 
-// We just set the defualts here in case the user does not specify them.
-// This avoids the controller trying to update itself right after creation.
+// customSetDefaults sets the default fields for DirectInternetAccess and RootAccess.
```
Comment (mbaijal) [inline pkg/resource/notebook_instance/custom_set_delta.go:7]: thank you!

File: pkg/resource/notebook_instance/custom_set_input_update.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	"github.com/aws/aws-sdk-go/aws"
+	svcsdk "github.com/aws/aws-sdk-go/service/sagemaker"
+)
+
+// handleUpdateOnlyParameters sets Disassociate<field> to true if corresponding value in desired is nil(ie not included in the spec)
+// and if the corresponding value in the latest spec is not nil (ie contains some value)
+// Ex if LifecycleConfigName was "a" in the latest spec but nil in the desired spec, DisassociateLifecycleConfig is set to true
```
Comment (mbaijal) [inline pkg/resource/notebook_instance/custom_set_input_update.go:11]: This one is a good example of a descriptive comment on a function.

File: pkg/resource/notebook_instance/hooks.go [hunk 10]
```diff
@@ -37,8 +41,8 @@ func (rm *resourceManager) requeueUntilCanModify(
 	ctx context.Context,
 	r *resource,
 ) error {
-	latestStatus := r.ko.Status.NotebookInstanceStatus
-	return svccommon.RequeueIfModifying(latestStatus, &resourceName, &modifyingStatuses)
+	notebookStatus := r.ko.Status.NotebookInstanceStatus
```
Comment (mbaijal) [inline pkg/resource/notebook_instance/hooks.go:44]: why the change here from latestStatus to notebookStatus ?

File: templates/notebook_instance/sdk_delete_pre_build_request.go.tpl [hunk 11]
```diff
@@ -1,3 +1,4 @@
+//This will avoid exponential backoff
```
Comment (mbaijal) [inline templates/notebook_instance/sdk_delete_pre_build_request.go.tpl:1]: nit: space after the `//`

File: templates/notebook_instance/sdk_update_pre_build_request.go.tpl [hunk 12]
```diff
@@ -0,0 +1,16 @@
+if err = rm.requeueUntilCanModify(ctx, latest); err != nil {
+	return latest, err
+}
+latestStatus := latest.ko.Status.NotebookInstanceStatus
+// The Notebook Instance is stopped and the StoppedByController status is
```
Comment (mbaijal) [inline templates/notebook_instance/sdk_update_pre_build_request.go.tpl:5]: nit: field name is incorrect

File: templates/notebook_instance/sdk_update_pre_build_request.go.tpl [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+// set to UpdatePending.
+if latestStatus != nil &&
+ *latestStatus == svcsdk.NotebookInstanceStatusInService {
+	if err := rm.stopNotebookInstance(latest); err != nil {
+		return latest, err
+	} else {
+		//TODO: Take this out if the runtime supports updating annotations if an error is returned and use annotations for this.
```
Comment (mbaijal) [inline templates/notebook_instance/sdk_update_pre_build_request.go.tpl:12]: Could you please reword this TODO.

File: templates/notebook_instance/sdk_update_pre_build_request.go.tpl [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+		latest.ko.Status.StoppedByControllerMETA = aws.String("UpdatePending")
+	  return latest, requeueWaitWhileStopping
```
Comment (mbaijal) [inline templates/notebook_instance/sdk_update_pre_build_request.go.tpl:14]: This else block seems to be malformed (or is just the indentation on github ?)

File: test/e2e/tests/test_notebook_instance.py [hunk 15]
```diff
@@ -146,7 +150,65 @@ def testCreateAndDelete(self, notebook_instance):
         )
         assert k8s.wait_on_condition(reference, "ACK.ResourceSynced", "True")
 
+    def update_notebook_test(self, notebook_instance):
+        (reference, resource, spec) = notebook_instance
```
Comment (mbaijal) [inline test/e2e/tests/test_notebook_instance.py:154]: This test is now creating multiple instances of the notebook, have you verified that - 
1. All 3 of them are getting properly cleaned up ?
2. Is this increasing the test run time by a lot ?

You should be able to use the same notebook instance to run create, update and delete. Just change the scope of the pytest fixture `notebook_instance` to module and check if that is enough.

File: test/e2e/tests/test_notebook_instance.py [hunk 16]
```diff
@@ -146,7 +150,65 @@ def testCreateAndDelete(self, notebook_instance):
...
+        notebook_instance_name = resource["spec"].get("notebookInstanceName", None)
+        volumeSizeInGB = 7
+
+        # Update test
+        spec["spec"]["volumeSizeInGB"] = volumeSizeInGB
+        # TODO: Use del spec["spec"]["defaultCodeRepository"] instead once ack.testinfra supports replacement.
+        # Patch only supports updating spec fields instead of fully getting rid of them.
+        spec["spec"]["defaultCodeRepository"] = None
+        k8s.patch_custom_resource(reference, spec)
```
Comment (mbaijal) [inline test/e2e/tests/test_notebook_instance.py:163]: could we also add an update test where the field did not exist before but is being added in the update ?

File: config/crd/bases/sagemaker.services.k8s.aws_notebookinstances.yaml [hunk 17]
```diff
@@ -224,6 +224,10 @@ spec:
               notebookInstanceStatus:
                 description: The status of the notebook instance.
                 type: string
+              stoppedByControllerMETA:
+                description: The URL that you use to connect to the Jupyter notebook
```
Comment (mbaijal) [inline config/crd/bases/sagemaker.services.k8s.aws_notebookinstances.yaml:228]: description is incorrect

**Review by mbaijal — CHANGES_REQUESTED** [review]

File: templates/notebook_instance/sdk_read_one_pre_set_output.go.tpl [hunk 18]
```diff
@@ -0,0 +1,6 @@
+//NotebookInstanceLifecycleConfigName does not get generated by the code generator.
+if resp.NotebookInstanceLifecycleConfigName != nil {
+	ko.Spec.LifecycleConfigName = resp.NotebookInstanceLifecycleConfigName
+} else {
```
[the remark this answers is not in the document]
Comment (ananth102) [reply templates/notebook_instance/sdk_read_one_pre_set_output.go.tpl:4]: It is, otherwise the user won't be able to update LifecycleConfigName.

File: generator.yaml (hunk 7 above)
Comment (ananth102) [reply generator.yaml:655]: Yes if you are referring to sdk_update_post_set_output.

File: pkg/resource/notebook_instance/hooks.go (hunk 10 above)
Comment (ananth102) [reply pkg/resource/notebook_instance/hooks.go:44]: I think I changed it in response to another PR comment.

File: templates/notebook_instance/sdk_read_one_pre_set_output.go.tpl [hunk 19]
```diff
@@ -0,0 +1,6 @@
+//NotebookInstanceLifecycleConfigName does not get generated by the code generator.
```
[the remark this answers is not in the document]
Comment (ananth102) [reply templates/notebook_instance/sdk_read_one_pre_set_output.go.tpl]: Yea, just moved it there.

File: pkg/resource/notebook_instance/hooks.go (hunk 2 above)
Comment (ananth102) [reply pkg/resource/notebook_instance/hooks.go]: I updated it to StoppedByControllerMetadata

File: pkg/resource/notebook_instance/hooks.go (hunk 3 above)
Comment (ananth102) [reply pkg/resource/notebook_instance/hooks.go]: It will be tested in the unit tests.  startNotebookInstance rarely returns errors.

File: pkg/resource/notebook_instance/custom_set_input_update.go (hunk 9 above)
Comment (ananth102) [reply pkg/resource/notebook_instance/custom_set_input_update.go:11]: Thanks! :)

File: templates/notebook_instance/sdk_update_pre_build_request.go.tpl (hunk 13 above)
Comment (ananth102) [reply templates/notebook_instance/sdk_update_pre_build_request.go.tpl]: done

File: templates/notebook_instance/sdk_update_pre_build_request.go.tpl (hunk 14 above)
Comment (ananth102) [reply templates/notebook_instance/sdk_update_pre_build_request.go.tpl:14]: It's just a github thing, looks fine on vscode

File: test/e2e/tests/test_notebook_instance.py (hunk 16 above)
Comment (ananth102) [reply test/e2e/tests/test_notebook_instance.py:165]: Added one for additionalCodeRepositories.

File: config/crd/bases/sagemaker.services.k8s.aws_notebookinstances.yaml (hunk 17 above)
Comment (ananth102) [reply config/crd/bases/sagemaker.services.k8s.aws_notebookinstances.yaml:228]: It's because the file is auto-generated, another reason to use annotations.

File: test/e2e/tests/test_notebook_instance.py (hunk 15 above)
Comment (ananth102) [reply test/e2e/tests/test_notebook_instance.py:154]: Don't think it was creating 3 notebooks before but I still added the module scope just to be safe. The Notebook is properly cleaned up here https://github.com/aws-controllers-k8s/sagemaker-controller/blob/0d63e02e1aa5b32ec62811bf679d6e6eea429409/test/e2e/tests/test_notebook_instance.py#L54.

**Comment by ananth102** [comment]
> > TODO: In test/e2e/tests/test_notebook_instance.py, we should use replace_custom_resource when test-infra supports it.
> 
> Is there an issue already tracking this request ?

@mbaijal  Not yet.

File: apis/v1alpha1/generator.yaml [hunk 20]
```diff
@@ -663,6 +671,11 @@ resources:
         from:
           operation: DescribeNotebookInstance
           path: Url
+      stoppedByControllerMETA:
+        is_read_only: true
+        from:
+          operation: DescribeNotebookInstance
+          path: Url
```
Comment (akartsky) [inline apis/v1alpha1/generator.yaml:676]: this is weird 
can we not have the value as null instead ?

Comment (ananth102) [reply apis/v1alpha1/generator.yaml:676]: Not really since code generator needs the shape(string/map/array) the generate the status.

**Review by akartsky — APPROVED** [review]

**Review by mbaijal — APPROVED** [review]
thanks for addressing the comments. 
Could you make sure to creare a github issue for the pending task.

File: pkg/resource/notebook_instance/custom_set_input_update.go (hunk 9 above)
Comment (surajkota) [reply pkg/resource/notebook_instance/custom_set_input_update.go:11]: good

File: test/e2e/tests/test_notebook_instance.py (hunk 16 above)
Comment (surajkota) [reply test/e2e/tests/test_notebook_instance.py:172]: good!

**Review by surajkota — APPROVED** [review]
/approve
Nice work
